File: PathoFroh/src/main/java/com/patho/main/template/print/SendReport.java
package com.patho.main.template.print;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.stream.Collectors;

import com.patho.main.model.PDFContainer;
import com.patho.main.template.InitializeToken;
import com.patho.main.template.PrintDocument;
import com.patho.main.util.helper.HistoUtil;
import com.patho.main.util.notification.NotificationContainer;
import com.patho.main.util.pdf.PDFCreator;
import com.patho.main.util.print.LoadedPrintPDFBearer;

/**
 * patient, <br>
 * task, <br>
 * diagnosisRevisions. <br>
 * useMail, <br>
 * mails, <br>
 * useFax, <br>
 * faxes, <br>
 * useLetter, <br>
 * letters, <br>
 * usePhone,<br>
 * phonenumbers, <br>
 * reportDate
 * 
 * @author andi
 *
 */
public class SendReport extends PrintDocument {

	private boolean useMail;
	private List<NotificationContainer> mails;

	private boolean useFax;
	private List<NotificationContainer> faxes;

	private boolean useLetters;
	private List<NotificationContainer> letters;

	private boolean usePhone;
	private List<NotificationContainer> phonenumbers;

	public SendReport(PrintDocument document) {
		super(document);
		setAfterPDFCreationHook(true);
	}

	public void initilize(InitializeToken... token) {

		for (InitializeToken initializeToken : token) {
			switch (initializeToken.getKey()) {
			case "useMail":
				useMail = (boolean) initializeToken.getValue();
				break;
			case "mails":
				mails = (List<NotificationContainer>) initializeToken.getValue();
				break;
			case "useFax":
				useFax = (boolean) initializeToken.getValue();
				break;
			case "faxes":
				faxes = (List<NotificationContainer>) initializeToken.getValue();
				break;
			case "useLetter":
				useLetters = (boolean) initializeToken.getValue();
				break;
			case "letters":
				letters = (List<NotificationContainer>) initializeToken.getValue();
				break;
			case "usePhone":
				usePhone = (boolean) initializeToken.getValue();
				break;
			case "phonenumbers":
				phonenumbers = (List<NotificationContainer>) initializeToken.getValue();
				break;
			default:
				break;
			}
		}

		HashMap<String, Object> map = new HashMap<String, Object>();

		for (int i = 0; i < token.length; i++) {
			map.put(token[i].getKey(), token[i].getValue());
		}
		initilize(map);
	}

	public PDFContainer onAfterPDFCreation(PDFContainer container, PDFCreator creator) {
		List<LoadedPrintPDFBearer> attachPdf = new ArrayList<LoadedPrintPDFBearer>();

		attachPdf.add(new LoadedPrintPDFBearer(container));

//		if (useFax && HistoUtil.isNotNullOrEmpty(mails))
//			attachPdf.addAll(mails.stream().filter(p -> p.getPdf() != null).map(p -> new LoadedPrintPDFBearer(p))
//					.collect(Collectors.toList()));
//
//		if (useFax && HistoUtil.isNotNullOrEmpty(faxes))
//			attachPdf.addAll(faxes.stream().filter(p -> p.getPdf() != null).map(p -> new LoadedPrintPDFBearer(p.getPdf()))
//					.collect(Collectors.toList()));
//
//		if (useLetters && HistoUtil.isNotNullOrEmpty(letters))
//			attachPdf.addAll(letters.stream().filter(p -> p.getPdf() != null)
//					.map(p -> new LoadedPrintPDFBearer(p.getPdf())).collect(Collectors.toList()));
//
//		if (usePhone && HistoUtil.isNotNullOrEmpty(phonenumbers))
//			attachPdf.addAll(phonenumbers.stream().filter(p -> p.getPdf() != null)
//					.map(p -> new LoadedPrintPDFBearer(p.getPdf())).collect(Collectors.toList()));

		container = creator.mergePDFs(container, attachPdf);

		return container;
	}
}

File: PathoFroh/src/main/kotlin/com/patho/main/template/print/SendReport.kt
package com.patho.main.template.print

import com.patho.main.model.PDFContainer
import com.patho.main.template.PrintDocument
import com.patho.main.util.pdf.PDFCreator
import com.patho.main.util.print.LoadedPrintPDFBearer
import com.patho.main.util.report.ui.ReportIntentNotificationUIContainer
import org.apache.velocity.context.Context
import java.util.*

/**
 * DiagnosisReport <br>
 * Values to replace: <br>
 * patient, <br>
 * task, <br>
 * diagnosisRevisions. <br>
 * useMail, <br>
 * mails, <br>
 * useFax, <br>
 * faxes, <br>
 * useLetter, <br>
 * letters, <br>
 * usePhone,<br>
 * phonenumbers, <br>
 * reportDate
 */
class SendReport(printDocument: PrintDocument) : PrintDocument(printDocument) {

    private var useMail: Boolean = false
    private var mails: List<ReportIntentNotificationUIContainer> = listOf()

    private var useFax: Boolean = false
    private var faxes: List<ReportIntentNotificationUIContainer> = listOf()

    private var useLetters: Boolean = false
    private var letters: List<ReportIntentNotificationUIContainer> = listOf()

    private var usePhone: Boolean = false
    private var phonenumbers: List<ReportIntentNotificationUIContainer> = listOf()

    init {
        afterPDFCreationHook = true
    }

    override fun initialize(content: HashMap<String, Any?>): Pair<out PrintDocument, Context> {
        content.forEach { p ->
            when (p.key) {
                "useMail" -> useMail = p.value as Boolean
                "mails" -> mails = p as List<ReportIntentNotificationUIContainer>
                "useFax" -> useFax = p.value as Boolean
                "faxes" -> faxes = p as List<ReportIntentNotificationUIContainer>
                "useLetter" -> useLetters = p.value as Boolean
                "letters" -> letters = p as List<ReportIntentNotificationUIContainer>
                "usePhone" -> usePhone = p.value as Boolean
                "phonenumbers" -> phonenumbers = p as List<ReportIntentNotificationUIContainer>
                else -> {
                }
            }
        }
        return super.initialize(content)
    }

    override fun onAfterPDFCreation(container: PDFContainer, creator: PDFCreator): PDFContainer {
        var container = container
        val attachPdf = ArrayList<LoadedPrintPDFBearer>()

        attachPdf.add(LoadedPrintPDFBearer(container))

        if (useMail && mails.isEmpty())
            attachPdf.addAll(mails.filter { p -> p.pdf != null }.map { LoadedPrintPDFBearer(it.pdf!!) })

        if (useFax && faxes.isEmpty())
            attachPdf.addAll(faxes.filter { p -> p.pdf != null }.map { LoadedPrintPDFBearer(it.pdf!!) })

        if (useLetters && letters.isEmpty())
            attachPdf.addAll(letters.filter { p -> p.pdf != null }.map { LoadedPrintPDFBearer(it.pdf!!) })

        if (usePhone && phonenumbers.isEmpty())
            attachPdf.addAll(phonenumbers.filter { p -> p.pdf != null }.map { LoadedPrintPDFBearer(it.pdf!!) })

        container = creator.mergePDFs(container, attachPdf)

        return container
    }
}